File: src/sec_af/agents/hunt/injection.py
from __future__ import annotations

import shutil
import tempfile
from pathlib import Path
from typing import TYPE_CHECKING, Protocol

from sec_af.agents._utils import extract_harness_result
from sec_af.context import recon_context_for_injection
from sec_af.schemas.hunt import HuntResult

if TYPE_CHECKING:
    from sec_af.schemas.recon import ReconResult


class HarnessCapable(Protocol):
    async def harness(
        self, prompt: str, *, schema: object = None, cwd: str | None = None, **kwargs: object
    ) -> object: ...


PROMPT_PATH = Path(__file__).resolve().parents[4] / "prompts" / "hunt" / "injection.txt"


async def run_injection_hunter(
    app: HarnessCapable,
    repo_path: str,
    recon_result: ReconResult,
    depth: str,
    max_files_without_signal: int = 30,
) -> HuntResult:
    prompt_template = PROMPT_PATH.read_text(encoding="utf-8")
    prompt = (
        prompt_template.replace("{{RECON_CONTEXT}}", recon_context_for_injection(recon_result))
        + "\n\nCONTEXT:\n"
        + f"- Repository path: {repo_path}\n"
        + f"- Depth profile: {depth}\n"
        + "- Early stop rule: if you inspect "
        + f"{max_files_without_signal} files without credible signal, "
        + "stop and return empty findings.\n"
        + "- Focus on RECON entry points and data flows as primary source-to-sink paths.\n"
        + "- Explore the codebase, trace data flows from sources to sinks, and identify injection points.\n"
        + "- Take multiple turns to build findings incrementally and write final JSON only when complete."
    )
    agent_name = "hunt-injection"
    harness_cwd = tempfile.mkdtemp(prefix=f"secaf-{agent_name}-")
    try:
        result = await app.harness(prompt=prompt, schema=HuntResult, cwd=harness_cwd, project_dir=repo_path)
        return extract_harness_result(result, HuntResult, "Injection hunter")
    finally:
        shutil.rmtree(harness_cwd, ignore_errors=True)

File: go/internal/agents/hunt/injection.go
package hunt

// Ports src/sec_af/agents/hunt/injection.py.

import (
	"context"
	"strconv"
	"strings"

	"github.com/Agent-Field/sec-af/go/internal/appx"
	"github.com/Agent-Field/sec-af/go/internal/prompts"
	"github.com/Agent-Field/sec-af/go/internal/recontext"
	"github.com/Agent-Field/sec-af/go/internal/schemas"
)

const injectionPromptPath = "hunt/injection.txt"

// injectionScanPrompt builds the exact prompt run_injection_hunter sends, and
// returns the recon context it embedded — the hunter reuses that same string
// for the enrichment step, so it is computed once here.
//
// earlyStop is the pre-rendered text for the early-stop rule. Python
// interpolates `max_files_without_signal` there with an f-string and never
// checks its type; see the package doc for why that matters.
func injectionScanPrompt(repoPath string, recon schemas.ReconResult, depth, earlyStop string) (scanPrompt, reconContext string) {
	reconContext = recontext.ReconContextForInjection(recon)
	template := prompts.MustLoad(injectionPromptPath)
	template = strings.ReplaceAll(template, "{{RECON_CONTEXT}}", reconContext)
	template = strings.ReplaceAll(template, "{{LANGUAGE_HINTS}}", recontext.LanguageHintsForContext(recon))
	template = strings.ReplaceAll(template, "{{FRAMEWORK_HINTS}}", recontext.FrameworkHintsForContext(recon))

	scanPrompt = template +
		"\n\nCONTEXT:\n" +
		"- Repository path: " + repoPath + "\n" +
		"- Depth profile: " + depth + "\n" +
		"- Early stop rule: if you inspect " +
		earlyStop + " files without credible signal, " +
		"stop and return empty findings.\n" +
		"- Focus on RECON entry points and data flows as primary source-to-sink paths.\n" +
		"- Explore the codebase, trace data flows from sources to sinks, and identify injection points.\n" +
		"- Take multiple turns to build findings incrementally and write final JSON only when complete."
	return scanPrompt, reconContext
}

// runInjectionHunter is the shared body; earlyStop is pre-rendered text so the
// hunt table can pass the value Python's argument cascade actually delivers.
func runInjectionHunter(
	ctx context.Context, app appx.Harnesser, repoPath string,
	recon schemas.ReconResult, depth, earlyStop string,
) (schemas.HuntResult, error) {
	scanPrompt, reconContext := injectionScanPrompt(repoPath, recon, depth, earlyStop)
	return runHunterBody(ctx, app, repoPath, hunterSpec{
		ScanPrompt:   scanPrompt,
		ReconContext: reconContext,
		FindingType:  "sast",
		Strategy:     "injection",
		// Python parity: `return HuntResult()` — strategies_run keeps its []
		// default rather than naming the strategy that just ran.
		EmptyStrategiesRun: nil,
	})
}

// RunInjectionHunter ports injection.py run_injection_hunter:
//
//	async def run_injection_hunter(app, repo_path, recon_result, depth,
//	                               max_files_without_signal: int = 30) -> HuntResult
func RunInjectionHunter(
	ctx context.Context, app appx.Harnesser, repoPath string,
	recon schemas.ReconResult, depth string, maxFilesWithoutSignal int,
) (schemas.HuntResult, error) {
	return runInjectionHunter(ctx, app, repoPath, recon, depth, strconv.Itoa(maxFilesWithoutSignal))
}
